docs/pipelines.md

```diff
@@ -574,6 +574,13 @@ After a `Pipeline` has executed the `PipelineRun` will be populated with the `Re
 emitted by the `Pipeline`. These will be written to the `PipelineRun's`
 `status.pipelineResults` field.
 
+A `Pipeline Result` is not emitted if any of the following are true:
+- a `TaskRun` referenced by the `Pipeline Result` failed
+- a `TaskRun` referenced by the `Pipeline Result` was skipped
+- a `TaskRun` referenced by the `Pipeline Result` didn't emit the referenced `Task Result`
+- the `Pipeline Result` uses a variable that doesn't point to a task in the Pipeline
+- the `Pipeline Result` uses a variable that doesn't point to a result of a task in the Pipeline
```

Comment:
@sbwsg continuation of #3497 - these all seem like errors to me

- the `Pipeline Result` uses a variable that doesn't point to a task in the Pipeline
- the `Pipeline Result` uses a variable that doesn't point to a result of a task in the Pipeline

If I'm understanding correctly, these both mean that the Pipeline was created incorrectly - we could catch these at validation time before we start actually executing the Pipeline

1. a `TaskRun` referenced by the `Pipeline Result` failed <-- this would fail the pipelinerun anyway
2. a `TaskRun` referenced by the `Pipeline Result` was skipped
3. a `TaskRun` referenced by the `Pipeline Result` didn't emit the referenced `Task Result`

These would all be at runtime - I could imagine that someone might be annoyed if their entire Pipeline ran successfully but one of the above caused a pipeline level result not to be set - however I think that could be addressed by a combo of:

1. https://github.com/tektoncd/pipeline/issues/3497 <-- case (3) would fail after the Task ran
2. https://github.com/tektoncd/community/pull/240 <-- we could include this use case if we want to be able to support defaults in the case of (2) (and maybe 3 if we let a Task specify a default result?)

Reply:
Agreed on all points.  Validation should catch incorrect variable usage and Tasks that don't emit any described results should fail, causing the Pipeline to also fail.

Regarding the validation, we do have this - it runs during DAG scheduling, so when we decide what to run next we find an invalid result reference and error the PipelineRun with `InvalidTaskResultReference`.

> we could catch these at validation time before we start actually executing the Pipeline

We could push this earlier and validate all Task result references before execution begins but we would need to fetch and inspect every `Task` referenced by a result variable before starting the Pipeline.  Otherwise we don't know what the correct result names are.  So it's feasible and do-able.  I think we need to decide if that's preferable vs whatever time it costs to fetch all the referenced Tasks and inspect their results before running the Pipeline.

Reading back I am realizing this morning that I was wrong in [my comment on #3497](https://github.com/tektoncd/pipeline/issues/3497#issuecomment-721850638). A PipelineRun was never put into a failed state just because a Task Result reference is invalid.  A PipelineRun is marked failed if it's unable to fetch the PipelineSpec needed to process Pipeline Results.  Otherwise the way it's worked before and after this change is that missing or invalid results don't update the final Success/Fail state of the PipelineRun.

Reply:
I've created #3499 to explore the possibility and costs of fetching and validating task result references before PipelineRun execution starts.

Reply:
I think I'm still a bit confused about both what the current behavior is and what the behavior will be after this PR - would it make sense to maybe make this section of the docs also indicate what behavior we expect in each of these casess?

Reply:
I agree and open to merge this since there is no functionality change compared to the existing behavior. And continue this discussion with #3499.

Reply:
Updated the doc with more detail about the accompanying PipelineRun state when these conditions are hit.